File: show/src/show/VerbNet.java
package show;

import java.io.File;
import java.io.FilenameFilter;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.Vector;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

public class VerbNet {

	private Vector<VerbMember> vms;
	private Map<String, Vector<VerbMember>> vbNetVerbToClassMap;

	private void parseXmlFile(String pathName) {
		// get the factory
		DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
		try {

			// Using factory get an instance of document builder
			DocumentBuilder db = dbf.newDocumentBuilder();
			// parse using builder to get DOM representation of the XML file
			Document dom = db.parse(pathName);
			// get the root element
			Element docElement = dom.getDocumentElement();
			// get a node list of elements
			NodeList nl = docElement.getElementsByTagName("MEMBERS");
			if (nl != null && nl.getLength() > 0) {
				for (int i = 0; i < nl.getLength(); i++) {
					Node parent = ((Element) nl.item(i)).getParentNode();
					NamedNodeMap nm = parent.getAttributes();
					Node n = nm.getNamedItem("ID");
					// System.out.print(n.getNodeValue()+"
					// length="+((Element)nl.item(i)).getElementsByTagName("MEMBER").getLength()+"\n");
					vms.add(new VerbMember(n.getNodeValue(), ((Element) nl.item(i)).getElementsByTagName("MEMBER"),
							parent.getChildNodes(), vbNetVerbToClassMap));
				}
			}
		} catch (ParserConfigurationException pce) {
			pce.printStackTrace();
		} catch (SAXException se) {
			se.printStackTrace();
		} catch (IOException ioe) {
			ioe.printStackTrace();
		}
	}

	public VerbNet() {
		File vnDir = new File("F:\\lp\\source\\lists\\VerbNet");
		FilenameFilter onlyXMLFiles = new FilenameFilter() {
			public boolean accept(File dir, String name) {
				return name.endsWith(".xml");
			}
		};
		vbNetVerbToClassMap = new HashMap<String, Vector<VerbMember>>();
		vms = new Vector<VerbMember>();
		for (String vnXMLFile : vnDir.list(onlyXMLFiles))
			parseXmlFile(vnDir.getAbsolutePath() + "\\" + vnXMLFile);
	}
	
	private  Vector<VerbMember> getVerbClasses(String baseVerb, String phrasalVerb) {
		Vector<VerbMember> vms = Show.vn.vbNetVerbToClassMap.get(baseVerb);
		if (phrasalVerb.length() > 0)
		{
			Vector<VerbMember> vmsParticiple = Show.vn.vbNetVerbToClassMap.get(phrasalVerb);
			if (vmsParticiple != null) 
				vms = vmsParticiple;
		}
		return vms;
	}

	private  boolean like(String str1,String str2)
	{
		int min=Math.min(str1.length(),str2.length());
		return str1.substring(0, min).equals(str2.substring(0,min));
	}
	
	public String getClassNames(String baseVerb) {
		String names = "";
		if (baseVerb.length() == 0)
			return names;
		Vector<VerbMember> vmsl = vbNetVerbToClassMap.get(baseVerb);
		if (vmsl != null) {
			for (VerbMember vm : vmsl)
				names += vm.getName() + " ";
		}
		return names;
	}
	
	public String getClassNames2(String baseVerb) {
		String names = "";
		if (vbNetVerbToClassMap.get(baseVerb) == null)
			System.out.println("Cannot find verb " + baseVerb);
		else			
			for (VerbMember vm : vbNetVerbToClassMap.get(baseVerb)) 
				names += vm.getNameWithKinds();
		return names;
	}
	
	public  boolean isVerbClass(String baseVerb, String phrasalVerb, String verbClass)
	{
		Vector <VerbMember> vms=getVerbClasses(baseVerb, phrasalVerb);
		if (vms!=null)
			for (VerbMember vm:vms)
				if (like(vm.getName(),verbClass)) return true;
		return false;
	}
}

File: Web/pyLPBackEnd/EclipseWorkspace/pyLPBackEnd/src/VerbNet.py
"""VerbNet.py - Load VerbNet XML classes into vbNetVerbToClassMap.

Overview:
    Walks F:\\lp\\source\\lists\\VerbNet\\*.xml, parses each VNCLASS
    and VNSUBCLASS into VerbMember rows, and indexes verbs by name.
    Lookup helpers return class-name strings or a boolean class test.

Pipeline position:
    Python-side VerbNet, parallel to the C++ VerbNet load during
    initialization. Used when the web backend needs Levin/VN classes.

Key entry points:
    - parse_xml_file(pathName) - one XML class file
    - __init__() - glob all XML under the hardcoded path
    - get_verb_classes / get_class_names / get_class_names_2 / is_verb_class
    - like(str1,str2) - prefix-equal-of-shorter-length compare

Notes / gotchas:
    Hardcoded F:\\lp\\source\\lists\\VerbNet and os.chdir.
    (fixed) get_verb_classes used phrasalVerb.length(), a Java-ism that
    raised AttributeError on a Python str; now len(phrasalVerb).
    (fixed) like() used str1.len()/str2.len() (AttributeError on str)
    and sliced with the `min` builtin instead of the local `m`; now
    uses len(str1)/len(str2) and str2[0:m]. Neither get_verb_classes nor
    like()/is_verb_class() on this class is currently called anywhere
    in pyLPBackEnd (Source.py only calls get_class_names/
    get_class_names_2 on a VerbNet instance), so this was latent
    dead-code-until-called rather than an active bug.
"""
from VerbMember import VerbMember
import glob, os
import xml.etree.ElementTree as ET
import jsonpickle
from json import JSONEncoder

class VerbNet:
    
    # Parse one VerbNet XML file: root VNCLASS plus each VNSUBCLASS
    # become a VerbMember appended to self.vms (and the verb map).
    def parse_xml_file(self, pathName):
        # get the factory
        tree = ET.parse(pathName)
        # get the root element
        root = tree.getroot()
        # get a node list of elements named MEMBERS
        if root != None:
            classId = root.get("ID")
            self.vms.append(VerbMember(classId, root.findall("./MEMBERS/MEMBER"), root.findall('*'),self.vbNetVerbToClassMap))
        nl = root.findall(".//VNSUBCLASS")
        if nl != None:
            for m in nl:
                classId = m.get("ID")
                self.vms.append(VerbMember(classId, m.findall("./MEMBERS/MEMBER"),m.findall('*'),self.vbNetVerbToClassMap))

    # chdir to the VerbNet XML directory and parse every *.xml.
    def __init__(self):
        self.vbNetVerbToClassMap = {}
        self.vms = []
        os.chdir("F:\\lp\\source\\lists\\VerbNet")
        for file in glob.glob("*.xml"):
            self.parse_xml_file(file);
            
    # Return the VerbMember list for baseVerb, or for phrasalVerb if
    # that key is present.
    def get_verb_classes(self, baseVerb, phrasalVerb):
        vms = self.vbNetVerbToClassMap.get(baseVerb)
        if len(phrasalVerb) > 0:
            vmsParticiple = self.vbNetVerbToClassMap.get(phrasalVerb)
            if vmsParticiple is not None:
                vms = vmsParticiple
        return vms

    # Prefix-equal of the shorter length.
    def like(self, str1,str2):
        m = min(len(str1),len(str2));
        return str1[0:m] == str2[0:m]
    
    # Space-joined vm.get_name() for every class of baseVerb, or "".
    def get_class_names(self, baseVerb):
        names = ""
        if len(baseVerb) == 0:
            return names;
        vmsl = self.vbNetVerbToClassMap.get(baseVerb)
        if vmsl is not None:
            for vm in vmsl:
                names += vm.get_name() + " "
        return names
    
    # Like get_class_names but uses get_name_with_kinds().
    def get_class_names_2(self, baseVerb):
        names = "";
        vml = self.vbNetVerbToClassMap.get(baseVerb)
        if vml is not None:
            for vm in vml: 
                names += vm.get_name_with_kinds()
        return names
    
    # True if any VN class of baseVerb/phrasalVerb is like(verbClass).
    def is_verb_class(self, baseVerb, phrasalVerb, verbClass):
        vms = self.get_verb_classes(baseVerb, phrasalVerb)
        if vms is not None:
            for vm in vms:
                if self.like(vm.get_name(),verbClass): 
                    return True
        return False
